Make checkpoint_sequential work with multiple arguments

andersenchen opened the pull request:
Summary:
In this commit, we make checkpoint_sequential work for models with multiple tensor inputs. Previously, it only processed the first tensor and ignored the rest.

We introduce a new test in test/test_utils.py that replicates the issue referenced in this [GitHub issue](https://github.com/pytorch/pytorch/issues/11093), and we make sure that the test passes by changing the behavior of checkpoint_sequential to process all input tensors.

Differential Revision: D13144672

ezyang commented:
Python lint failure! Make sure to fix it.

ezyang commented on torch/utils/checkpoint.py:
```diff
@@ -129,10 +129,15 @@ def checkpoint_sequential(functions, segments, *inputs):
-            input = inputs[0]
+        def forward(input, *additional_inputs):
+            if not isinstance(input, tuple):
+                input = (input,)
```
The special handling of the first argument to `forward` specifically seems a bit weird. Can you explain which case this makes work out?

andersenchen replied:
Agreed - the original intention was for the emptiness of `additional_inputs` to determine whether we had multiple arguments or not (i.e. whether we had multiple tensors as input), and then toggle function behavior based on that. I've cleaned up the code so that just one argument, inputs, needs to be passed in; we now determine whether we have multiple tensors as input and toggle based on that by looking at `isinstance(inputs, tuple)`.

ezyang commented on torch/utils/checkpoint.py:
```diff
@@ -130,10 +130,12 @@ def checkpoint_sequential(functions, segments, *inputs):
+                    inputs = functions[j](*inputs)
+                else:
+                    inputs = functions[j](inputs)
+            return inputs
```
This looks a lot more like what I expected :)

ezyang reviewed (approved):
LGTM. If you want to improve the patch more, I'd try to reduce code duplication in the test, but as it stands I think it is just barely tolerable.

apaszke reviewed (dismissed):
What if one of the functions doesn't expect us to unwrap the tuple? There's really no way to tell if a function expects a single argument or not

andersenchen commented:
@ezyang Agreed; the code that checks checkpointed models against non-checkpointed models is used repeatedly in each of these two tests (`test_checkpoint_module_list_multiple_args` and `test_checkpoint`, so it makes sense to pull it out. Will do before closing this out.

andersenchen commented:
Hey @apaszke ! Yeah, one of the assumptions here is that functions passed into `checkpoint_sequential` don't expect tuples as single arguments as input.

In an earlier draft I had this function signature, which allowed us to differentiate between a tuple-as-a-single-arg and a tuple representing multiple args _in the first layer only_: `def forward(input, *additional_inputs):`

But I don't have a surefire way of doing this for subsequent layers; I can give this some more thought, but right now the best thing I can think of is passing passing args wrapped first and then checking for a TypeError. I have a feeling that this approach would get ugly pretty quick. Any thoughts @ezyang @apaszke ?

andersenchen commented:
1) @ezyang I refactored tests for checkpoint_sequential per your recommendation.

2) @ezyang @apaszke if there's no neat technical solution (e.g. there's really no way to tell if a user really just wants to pass in a tuple as a single arg), we could perhaps put a warning in the docs or even revisit how we want to deal with checkpoint_sequential and multiple args.

ezyang commented:
@apaszke:

> What if one of the functions doesn't expect us to unwrap the tuple? There's really no way to tell if a function expects a single argument or not

Well, first, the original was even more busted, because it just straight up dropped all of the arguments past the first one, without saying anything to the user. So if nothing else, we should accept *that* change.

OK, but maybe that's not the case you're worried about: you're worried about a forward returning a tuple, which is unwrapped. But this treatment is consistent with how the rest of the existing implementation of `checkpoint_sequential` works:

```
    for start in range(0, segment_size * (segments - 1), segment_size):
        end = start + segment_size - 1
        inputs = checkpoint(run_function(start, end, functions), *inputs)
        if not isinstance(inputs, tuple):
            inputs = (inputs,)
    return run_function(end + 1, len(functions) - 1, functions)(*inputs)
```

Another tuple check!

OK, but maybe what you're really saying is, all of this is broken, and we should redesign the API to have some sort of predictable behavior with regards to tuple. But the behavior seems pretty predictable to me: if your function returns a tuple, it will be interpreted as an argument list for the next function. If you actually wanted to return a tuple and not explode, return a tuple wrapped in a tuple.

So I don't see any reason why we shouldn't take this change.

andersenchen commented:
Hey @apaszke ! Was just wondering if you had an update on this PR; I'm in Bootcamp right now so my time is a bit bounded :)

Commit:
Make checkpoint_sequential work with multiple arguments (#14278)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/14278

In this commit, we make checkpoint_sequential work for models with multiple tensor inputs. Previously, it only processed the first tensor and ignored the rest.

We introduce a new test in test/test_utils.py that replicates the issue referenced in this [GitHub issue](https://github.com/pytorch/pytorch/issues/11093), and we make sure that the test passes by changing the behavior of checkpoint_sequential to process all input tensors.

Reviewed By: ezyang

Differential Revision: D13144672

fbshipit-source-id: 0100afa5e4dd06fa095ed2a7f59271b7ab466405

sublee commented:
I reported an inconsistency brought by this PR at #19260. `*args` of a sequential module is still illegal in `nn.Sequential` but it's legal under `checkpoint_seqeuntial`.

```python
import torch
import torch.utils.checkpoint

class Two(torch.nn.Module):
    def forward(self, a, b):
        return a, b
model = torch.nn.Sequential(Two())

a = torch.rand(1)
b = torch.rand(1)

torch.utils.checkpoint.checkpoint_sequential(model, 1, a, b)  # OK
model(a, b)  # TypeError!
# TypeError: forward() takes 2 positional arguments but 3 were given
```

I believe both `nn.Sequential` and `checkpoint_sequential` should provide the same interface. So a failed `nn.Sequential` model should fail under `checkpoint_sequential` too. Also, a succeeded `nn.Sequential` model should succeed under `checkpoint_sequential` too.
